cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh

```diff
@@ -45,11 +46,30 @@ namespace detail {
 
 namespace intersection_functors {
 
+/** @brief Functor to compute the updated offset buffer after `remove_if` operation.
+ *
+ *  Given the `i`th row in the `geometry_collection_offset`, find the number of all
+ *  previous removed geometry from `reduced_values` by searching `reduced_keys`.
+ *
+ *  Example:
+ *  i:                             0  1  2  3  4  5  6  7
+ *  Geometry Collection Offset:    0  0  0  1  3  4  4  4
+ *  Reduced Keys:                  3  4  5
+ *  j:                            -1 -1 -1  0  1  2  2  2
+ *  Reduced Values:                0  2  2                  (Removed 2 geometries)
+ *  Should subtract:               /  /  /  0  2  2  2  2
+ *  Result:                        0  0  0  0  1  2  2  2
```

Comment:
I get everything up to the `j:` line. But after that I'm trying to follow what these last 3 lines mean, and what "removed 2 geometries" means.  The code at this point is: 

```
if (j < 0) return offset;

return offset - reduced_values_begin[j];
```

But what is `offset` in the example above? And how does this code correspond with the last 3 lines of the example?

Reply:
`offset` is `GeometryCollectionOffset[i]` in the example. 

`GeometryCollectionOffset[i]` should subtract the total number of removed geometries in all preceding lists as a result of `remove_if`. This number is `reduced_value[j]`. `0 2 2` means:
```
0 geometries are removed for the first non empty list
2 geometries are removed for the first and the second nonempty list
2 geometries are removed for the first, second and third nonempty list
In total, 2 geometries are removed.
```

To me, the most difficult part to grasp is why `j` can be achieved by searching `reduced_key`.  Recall in the example:
```
Geometry Collection Offset:    0  0  0  1  3  4  4  4
reduced key: 3 4 5
```

The values in `reduced_key` represents the *end position of the indices* (index of index) for the 3 non-empty lists in the column. e.g. `GeometryCollectionOffset[3] == 1` means the first non empty list ends at position 1 of the column.
This makes `reduced_key` the boundary condition for partitioning the `GeometryCollectionOffset` array.

Reply:
Docs are updated.